File: src/Uno.UI/ts/Runtime/BrowserPointerInputSource.ts
﻿namespace Uno.UI.Runtime.Skia {
	//TODO import PointerDeviceType = Windows.Devices.Input.PointerDeviceType;
	import HtmlEventDispatchResult = Uno.UI.HtmlEventDispatchResult;

	export enum PointerDeviceType {
		Touch = 0,
		Pen = 1,
		Mouse = 2,
	}

	export enum HtmlPointerEvent {
		pointerover = 1,
		pointerleave = 1 << 1,
		pointerdown = 1 << 2,
		pointerup = 1 << 3,
		pointercancel = 1 << 4,

		// Optional pointer events
		pointermove = 1 << 5,
		lostpointercapture = 1 << 6,
		wheel = 1 << 7,
	}

	export class BrowserPointerInputSource {

		private static _exports: any;
		
		public static async initialize(inputSource: any): Promise<any> {

			const module = <any>window.Module;
			if (BrowserPointerInputSource._exports == undefined
				&& module.getAssemblyExports !== undefined) {
					
				const browserExports = (await module.getAssemblyExports("Uno.UI"));

				BrowserPointerInputSource._exports = browserExports.Uno.UI.Runtime.BrowserPointerInputSource;
			}

			return new BrowserPointerInputSource(inputSource);
		}

		public static setPointerCapture(pointerId: number): void {
			document.body.setPointerCapture(pointerId);
		}

		public static releasePointerCapture(pointerId: number): void {
			document.body.releasePointerCapture(pointerId);
		}

		private _source: any;
		private _bootTime: Number;

		private constructor(manageSource: any) {
			this._bootTime = Date.now() - performance.now();
			this._source = manageSource;

			BrowserPointerInputSource._exports.OnInitialized(manageSource, this._bootTime);
			this.subscribePointerEvents(); // Subscribe only after the managed initialization is done
		}

		private subscribePointerEvents() {
			const element = document.body;

			element.addEventListener("pointerover", this.onPointerEventReceived.bind(this), { capture: false });
			element.addEventListener("pointerleave", this.onPointerEventReceived.bind(this), { capture: false });
			element.addEventListener("pointerdown", this.onPointerEventReceived.bind(this), { capture: false });
			element.addEventListener("pointerup", this.onPointerEventReceived.bind(this), { capture: false });
			element.addEventListener("pointercancel", this.onPointerEventReceived.bind(this), { capture: false });
			element.addEventListener("pointermove", this.onPointerEventReceived.bind(this), { capture: false });
			element.addEventListener("wheel", this.onPointerEventReceived.bind(this), { capture: false });
		}

		private onPointerEventReceived(evt: PointerEvent): void {
			const event = BrowserPointerInputSource.toHtmlPointerEvent(evt.type);

			let pointerId: number, pointerType: PointerDeviceType, pressure: number;
			let wheelDeltaX: number, wheelDeltaY: number;
			if (evt instanceof WheelEvent) {
				pointerId = (evt as any).mozInputSource ? 0 : 1; // Try to match the mouse pointer ID 0 for FF, 1 for others
				pointerType = PointerDeviceType.Mouse;
				pressure = 0.5; // like WinUI
				wheelDeltaX = evt.deltaX;
				wheelDeltaY = evt.deltaY;

				switch (evt.deltaMode) {
					case WheelEvent.DOM_DELTA_LINE: // Actually this is supported only by FF
						const lineSize = BrowserPointerInputSource.wheelLineSize;
						wheelDeltaX *= lineSize;
						wheelDeltaY *= lineSize;
						break;
					case WheelEvent.DOM_DELTA_PAGE:
						wheelDeltaX *= document.documentElement.clientWidth;
						wheelDeltaY *= document.documentElement.clientHeight;
						break;
				}
			} else {
				pointerId = evt.pointerId;
				pointerType = BrowserPointerInputSource.toPointerDeviceType(evt.pointerType);
				pressure = evt.pressure;
				wheelDeltaX = 0;
				wheelDeltaY = 0;
			}

			const result = BrowserPointerInputSource._exports.OnNativeEvent(
				this._source,
				event, //byte @event, // ONE of NativePointerEvent
				evt.timeStamp, //double timestamp,
				pointerType, //int deviceType, // ONE of _PointerDeviceType
				pointerId, //double pointerId, // Warning: This is a Number in JS, and it might be negative on safari for iOS
				evt.clientX, //double x,
				evt.clientY, //double y,
				evt.ctrlKey, //bool ctrl,
				evt.shiftKey, //bool shift,
				evt.buttons, //int buttons,
				evt.button, //int buttonUpdate,
				pressure, //double pressure,
				wheelDeltaX, //double wheelDeltaX,
				wheelDeltaY, //double wheelDeltaY,
				evt.relatedTarget !== null //bool hasRelatedTarget)
			);
			// This is uesless with root pointer management
			//if (result & HtmlEventDispatchResult.StopPropagation) {
			//	evt.stopPropagation();
			//}
			if (result & HtmlEventDispatchResult.PreventDefault) {
				evt.preventDefault();
			}
		}

		//#region WheelLineSize
		private static _wheelLineSize: number = undefined;
		private static get wheelLineSize(): number {
			// In web browsers, scroll might happen by pixels, line or page.
			// But WinUI works only with pixels, so we have to convert it before send the value to the managed code.
			// The issue is that there is no easy way get the "size of a line", instead we have to determine the CSS "line-height"
			// defined in the browser settings. 
			// https://stackoverflow.com/questions/20110224/what-is-the-height-of-a-line-in-a-wheel-event-deltamode-dom-delta-line
			if (this._wheelLineSize == undefined) {
				const el = document.createElement("div");
				el.style.fontSize = "initial";
				el.style.display = "none";
				document.body.appendChild(el);
				const fontSize = window.getComputedStyle(el).fontSize;
				document.body.removeChild(el);

				this._wheelLineSize = fontSize ? parseInt(fontSize) : 16; /* 16 = The current common default font size */

				// Based on observations, even if the event reports 3 lines (the settings of windows),
				// the browser will actually scroll of about 6 lines of text.
				this._wheelLineSize *= 2.0;
			}

			return this._wheelLineSize;
		}
		//#endregion

		//#region Helpers
		private static toHtmlPointerEvent(eventName: string): HtmlPointerEvent {
			switch (eventName) {
				case "pointerover":
					return HtmlPointerEvent.pointerover;
				case "pointerleave":
					return HtmlPointerEvent.pointerleave;
				case "pointerdown"	 :
					return HtmlPointerEvent.pointerdown;
				case "pointerup"	 :
					return HtmlPointerEvent.pointerup;
				case "pointercancel" :
					return HtmlPointerEvent.pointercancel;
				case "pointermove"	 :
					return HtmlPointerEvent.pointermove;
				case "wheel":
					return HtmlPointerEvent.wheel;
				default:
					return undefined;
			}
		}

		private static toPointerDeviceType(type: string): PointerDeviceType {
			switch (type) {
				case "touch":
					return PointerDeviceType.Touch;
				case "pen":
					// Note: As of 2019-11-28, once pen pressed events pressed/move/released are reported as TOUCH on Firefox
					//		 https://bugzilla.mozilla.org/show_bug.cgi?id=1449660
					return PointerDeviceType.Pen;
				case "mouse":
				default:
					return PointerDeviceType.Mouse;
			}
		}
		//#endregion
	}
}

File: src/Uno.UI.Runtime.Skia.WebAssembly.Browser/Devices/Input/BrowserPointerInputSource.cs
using System;
using Windows.Devices.Input;
using Windows.Foundation;
using Windows.UI.Core;
using Windows.UI.Input;
using Microsoft.UI.Xaml.Controls;
using static Windows.UI.Input.PointerUpdateKind;
using Uno.Foundation.Logging;
using System.Runtime.InteropServices.JavaScript;

using _PointerIdentifierPool = Windows.Devices.Input.PointerIdentifierPool; // internal type (should be in Uno namespace)
using _PointerIdentifier = Windows.Devices.Input.PointerIdentifier; // internal type (should be in Uno namespace)
using System.Runtime.InteropServices;
using Windows.System;
using Uno.UI.Dispatching;
using Uno.UI.Xaml;

namespace Uno.UI.Runtime.Skia;

internal unsafe partial class BrowserPointerInputSource : IUnoCorePointerInputSource
{
	private static readonly Logger _log = typeof(BrowserPointerInputSource).Log();
	private static readonly Action<string>? _trace = _log.IsTraceEnabled() ? _log.Trace : null;

	private ulong _bootTime;
	private PointerPoint? _lastPoint;

#pragma warning disable CS0067 // Some event are not raised on skia browser ... yet!
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerCaptureLost;
#pragma warning restore CS0067 // Some event are not raised on skia browser ... yet!
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerEntered;
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerExited;
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerMoved;
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerPressed;
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerReleased;
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerWheelChanged;
	public event TypedEventHandler<object, Windows.UI.Core.PointerEventArgs>? PointerCancelled; // Uno Only

	public BrowserPointerInputSource()
	{
		_trace?.Invoke("Initializing BrowserPointerInputSource");

		Initialize(this);
	}

	[JSImport("globalThis.Uno.UI.Runtime.Skia.BrowserPointerInputSource.initialize")]
	private static partial void Initialize([JSMarshalAs<JSType.Any>] object inputSource);

	[JSExport]
	private static void OnInitialized([JSMarshalAs<JSType.Any>] object inputSource, double bootTime)
	{
		((BrowserPointerInputSource)inputSource)._bootTime = (ulong)bootTime;

		_trace?.Invoke("Complete initialization of BrowserPointerInputSource, we are now ready to receive pointer events!");
	}

	[JSExport]
	[return: JSMarshalAs<JSType.Number>]
	private static int OnNativeEvent(
		[JSMarshalAs<JSType.Any>] object inputSource,
		byte @event, // ONE of NativePointerEvent
		double timestamp,
		int deviceType, // ONE of _PointerDeviceType
		double pointerId, // Warning: This is a Number in JS, and it might be negative on safari for iOS
		double x,
		double y,
		bool ctrl,
		bool shift,
		int buttons,
		int buttonUpdate,
		double pressure,
		double wheelDeltaX,
		double wheelDeltaY,
		bool hasRelatedTarget)
	{

		try
		{
			_trace?.Invoke($"Pointer evt={(HtmlPointerEvent)@event}|id={pointerId}|x={x}|y={x}|ctrl={ctrl}|shift={shift}|bts={buttons}|btUpdate={buttonUpdate}|type={(PointerDeviceType)deviceType}|ts={timestamp}|pres={pressure}|wheelX={wheelDeltaX}|wheelY={wheelDeltaY}|relTarget={hasRelatedTarget}");

			// Ensure that the async context is set properly, since we're raising
			// events from outside the dispatcher.
			using var syncContextScope = NativeDispatcher.Main.SynchronizationContext.Apply();

			var that = (BrowserPointerInputSource)inputSource;
			var evt = (HtmlPointerEvent)@event;
			var pointerType = (PointerDeviceType)deviceType;
			var pointerDevice = PointerDevice.For(pointerType);
			var pointerIdentifier = _PointerIdentifierPool.RentManaged(new _PointerIdentifier((PointerDeviceType)deviceType, (uint)pointerId));

			var frameId = ToFrameId(timestamp);
			var ts = that.ToTimestamp(timestamp);
			var isInContact = buttons != 0;
			var isInRange = GetIsInRange(@event, hasRelatedTarget, pointerType, isInContact);
			var keyModifiers = GetKeyModifiers(ctrl, shift);
			var position = new Point(x, y);

			var properties = GetProperties(pointerType, isInRange, (HtmlPointerButtonsState)buttons, (HtmlPointerButtonUpdate)buttonUpdate, wheel: (false, -wheelDeltaY), pressure);

			var point = new PointerPoint(frameId, ts, pointerDevice, pointerIdentifier.Id, position, position, isInContact, properties);
			var args = new PointerEventArgs(point, keyModifiers);

			that._lastPoint = point;

			switch (evt)
			{
				case HtmlPointerEvent.pointerover:
					that.PointerEntered?.Invoke(that, args);
					break;

				case HtmlPointerEvent.pointerout:
					that.PointerExited?.Invoke(that, args);
					break;

				case HtmlPointerEvent.pointerdown:
					that.PointerPressed?.Invoke(that, args);
					break;

				case HtmlPointerEvent.pointerup:
					//case HtmlPointerEvent.lostpointercapture: // if pointer is captured, we don't get a up, just a capture lost (with skia for wasm)
					that.PointerReleased?.Invoke(that, args);
					break;

				case HtmlPointerEvent.pointermove:
					that.PointerMoved?.Invoke(that, args);
					break;

				case HtmlPointerEvent.wheel:
					if (wheelDeltaY is not 0)
					{
						that.PointerWheelChanged?.Invoke(that, args);
					}

					if (wheelDeltaX is not 0)
					{
						properties = GetProperties(pointerType, isInRange, (HtmlPointerButtonsState)buttons, (HtmlPointerButtonUpdate)buttonUpdate, wheel: (true, wheelDeltaX), pressure);
						point = new PointerPoint(frameId, ts, pointerDevice, pointerIdentifier.Id, position, position, isInContact, properties);
						args = new PointerEventArgs(point, keyModifiers);

						that.PointerWheelChanged?.Invoke(that, args);
					}

					break;

				case HtmlPointerEvent.pointercancel:
					that.PointerCancelled?.Invoke(that, args);
					break;

				default:
					throw new ArgumentOutOfRangeException(nameof(@event), $"Unknown event ({@event}-{evt}).");
			}

			return (int)(args.Handled ? HtmlEventDispatchResult.PreventDefault : HtmlEventDispatchResult.Ok);
		}
		catch (Exception error)
		{
			if (_log.IsEnabled(LogLevel.Error))
			{
				_log.Error($"Failed to dispatch native pointer event: {error}");
			}

			return (int)Uno.UI.Xaml.HtmlEventDispatchResult.Ok; // TODO
		}
	}

	[NotImplemented] public bool HasCapture => false;

	private CoreCursor _pointerCursor = new(CoreCursorType.Arrow, 0);

	public CoreCursor PointerCursor
	{
		get => _pointerCursor;
		set
		{
			_pointerCursor = value;
			SetCursor(value.Type.ToCssCursor());
		}
	}

	[JSImport("globalThis.Uno.UI.Runtime.Skia.WebAssemblyWindowWrapper.setCursor")]
	private static partial void SetCursor(string cssCursor);

	public Point PointerPosition => _lastPoint?.Position ?? default;

	#region Captures
	public void SetPointerCapture()
	{
		if (_lastPoint is not null)
		{
			SetPointerCapture(_lastPoint.Pointer);
		}
	}
	public void SetPointerCapture(PointerIdentifier pointer)
	{
		if (_PointerIdentifierPool.TryGetNative(pointer, out var native))
		{
			SetPointerCaptureNative(native.Id);
		}
		else if (this.Log().IsEnabled(LogLevel.Warning))
		{
			this.Log().Warn($"Cannot capture pointer, could not find native pointer id for managed pointer id {pointer}");
		}
	}

	[JSImport("globalThis.Uno.UI.Runtime.Skia.BrowserPointerInputSource.setPointerCapture")]
	private static partial void SetPointerCaptureNative(double pointerId); // double as it might be negative on safari for iOS

	public void ReleasePointerCapture()
	{
		if (_lastPoint is not null)
		{
			ReleasePointerCapture(_lastPoint.Pointer);
		}
	}

	public void ReleasePointerCapture(PointerIdentifier pointer)
	{
		if (_PointerIdentifierPool.TryGetNative(pointer, out var native))
		{
			ReleasePointerCaptureNative(native.Id);
		}
		else if (this.Log().IsEnabled(LogLevel.Warning))
		{
			this.Log().Warn($"Cannot release pointer, could not find native pointer id for managed pointer id {pointer}");
		}
	}

	[JSImport("globalThis.Uno.UI.Runtime.Skia.BrowserPointerInputSource.releasePointerCapture")]
	private static partial void ReleasePointerCaptureNative(double pointerId); // double as it might be negative on safari for iOS
	#endregion

	#region Native to manage convertion helpers
	private static VirtualKeyModifiers GetKeyModifiers(bool ctrl, bool shift)
	{
		var keyModifiers = VirtualKeyModifiers.None;
		if (ctrl) keyModifiers |= VirtualKeyModifiers.Control;
		if (shift) keyModifiers |= VirtualKeyModifiers.Shift;
		return keyModifiers;
	}

	private static bool GetIsInRange(byte @event, bool hasRelatedTarget, PointerDeviceType pointerType, bool isInContact)
	{
		const int cancel = (int)HtmlPointerEvent.pointercancel;
		const int exitOrUp = (int)(HtmlPointerEvent.pointerout | HtmlPointerEvent.pointerup);

		var isInRange = true;
		if (@event is cancel)
		{
			isInRange = false;
		}
		else if ((@event & exitOrUp) != 0)
		{
			isInRange = pointerType switch
			{
				PointerDeviceType.Mouse => true, // Mouse is always in range (unless for 'cancel' eg. if it was unplugged from computer)
				PointerDeviceType.Touch => false, // If we get a pointer out for touch, it means that pointer left the screen (pt up - a.k.a. Implicit capture)
				PointerDeviceType.Pen => hasRelatedTarget, // If the relatedTarget is null it means pointer left the detection range ... but only for out event!
				_ => !isInContact // Safety!
			};
		}

		return isInRange;
	}

	private static PointerPointProperties GetProperties(
		PointerDeviceType deviceType,
		bool isInRange,
		HtmlPointerButtonsState buttons,
		HtmlPointerButtonUpdate buttonUpdate,
		(bool isHorizontalWheel, double delta) wheel,
		double pressure)
	{
		var props = new PointerPointProperties
		{
			IsPrimary = true,
			IsInRange = isInRange,
			IsLeftButtonPressed = buttons.HasFlag(HtmlPointerButtonsState.Left),
			IsMiddleButtonPressed = buttons.HasFlag(HtmlPointerButtonsState.Middle),
			IsRightButtonPressed = buttons.HasFlag(HtmlPointerButtonsState.Right),
			IsXButton1Pressed = buttons.HasFlag(HtmlPointerButtonsState.X1),
			IsXButton2Pressed = buttons.HasFlag(HtmlPointerButtonsState.X2),
			IsEraser = buttons.HasFlag(HtmlPointerButtonsState.Eraser),
			IsHorizontalMouseWheel = wheel.isHorizontalWheel,
			MouseWheelDelta = (int)wheel.delta
		};

		switch (deviceType)
		{
			// For touch and mouse, we keep the default pressure of .5, as WinUI

			case PointerDeviceType.Pen:
				// !!! WARNING !!! Here we have a slight different behavior compared to WinUI:
				// On WinUI we will get IsRightButtonPressed (with IsBarrelButtonPressed) only if the user is pressing
				// the barrel button when pen goes "in contact" (i.e. touches the screen), otherwise we will get
				// IsLeftButtonPressed and IsBarrelButtonPressed.
				// Here we set IsRightButtonPressed as soon as the barrel button was pressed, no matter
				// if the pen was already in contact or not.
				// This is acceptable since the UIElement pressed state is **per pointer** (not buttons of pointer)
				// and GestureRecognizer always checks that pressed buttons didn't changed for a single gesture.
				props.IsBarrelButtonPressed = props.IsRightButtonPressed;
				props.Pressure = (float)pressure;
				break;
		}

		// Needs to be computed only after the props is almost completed
		props.PointerUpdateKind = ToUpdateKind(buttonUpdate, props);

		return props;
	}

	internal static uint ToFrameId(double timestamp)
		// Known limitation: After 49 days, we will overflow the uint and frame IDs will restart at 0.
		=> (uint)(timestamp % uint.MaxValue);

	private ulong ToTimestamp(double timestamp)
		=> _bootTime + (ulong)(timestamp * 1000);

	private static PointerUpdateKind ToUpdateKind(HtmlPointerButtonUpdate update, PointerPointProperties props)
		=> update switch
		{
			HtmlPointerButtonUpdate.Left when props.IsLeftButtonPressed => PointerUpdateKind.LeftButtonPressed,
			HtmlPointerButtonUpdate.Left => PointerUpdateKind.LeftButtonReleased,
			HtmlPointerButtonUpdate.Middle when props.IsMiddleButtonPressed => PointerUpdateKind.MiddleButtonPressed,
			HtmlPointerButtonUpdate.Middle => PointerUpdateKind.MiddleButtonReleased,
			HtmlPointerButtonUpdate.Right when props.IsRightButtonPressed => PointerUpdateKind.RightButtonPressed,
			HtmlPointerButtonUpdate.Right => PointerUpdateKind.RightButtonReleased,
			HtmlPointerButtonUpdate.X1 when props.IsXButton1Pressed => PointerUpdateKind.XButton1Pressed,
			HtmlPointerButtonUpdate.X1 => PointerUpdateKind.XButton1Released,
			HtmlPointerButtonUpdate.X2 when props.IsXButton2Pressed => PointerUpdateKind.XButton1Pressed,
			HtmlPointerButtonUpdate.X2 => PointerUpdateKind.XButton1Released,
			_ => PointerUpdateKind.Other
		};

	private enum HtmlPointerEvent : byte
	{
		// WARNING: This enum has a corresponding version in TypeScript!

		// Minimal default pointer required to maintain state
		pointerover = 1,
		pointerout = 1 << 1,
		pointerdown = 1 << 2,
		pointerup = 1 << 3,
		pointercancel = 1 << 4,
		pointermove = 1 << 5, // Required since when pt is captured, isOver will be maintained by the moveWithOverCheck
		lostpointercapture = 1 << 6, // Required to get pointer up when pointer is captured

		// Optional pointer events
		wheel = 1 << 7,
	}

	[Flags]
	internal enum HtmlPointerButtonsState
	{
		// https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events#Determining_button_states

		None = 0,
		Left = 1,
		Middle = 4,
		Right = 2,
		X1 = 8,
		X2 = 16,
		Eraser = 32,
	}

	private enum HtmlPointerButtonUpdate
	{
		None = -1,
		Left = 0,
		Middle = 1,
		Right = 2,
		X1 = 3,
		X2 = 4,
		Eraser = 5
	}
	#endregion
}
